### ISSUE TITLE
segmentation on test data where there is no ground truth.

### ISSUE THREAD

**Issue by ninjakx** [qubvel/segmentation_models#263]
```
test_dataset = Dataset(
    x_test_dir, 
    y_test_dir, 
    classes=CLASSES, 
    augmentation=get_validation_augmentation(),
    preprocessing=get_preprocessing(preprocess_input),
)

test_dataloader = Dataloder(test_dataset, batch_size=1, shuffle=False)

```
this function takes 2 inputs- image and its mask.
I tried to make the dataset only accept input image but then how to use this line `preprocessing=get_preprocessing(preprocess_input)`

I have basically two MRI dataset. One is for training and another for testing(where I have to crop the tumour part). Training is done. Model is also working but how to predict on test data.
I predicted like this: 

```
for each in os.listdir('test/'):
    print(each)
    img = cv2.imread(os.path.join('test', each))
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = preprocessing_fn(img)
    img = img.transpose(2, 0, 1)#.astype('float32')
    img = img.astype('uint8')/255.0
    x = torch.from_numpy(np.expand_dims(img, 0))
    y = model.predict(x.to(DEVICE))
```

But the model is predicting in between (10^-6 to 10^-1) for the test dataset but when I tested it on the dataset which I used for training, validation. It gave me an output either 0 or 1.

**Comment by JordanMakesMaps** [comment]
It might be because you're rescaling the image:

`img = img.astype('uint8')/255.0`

even though you've already pre-processed it with `preprocessing_fn`.

I'm not sure what it does, so maybe it's not the problem, but I've found that usually when predictions are giving out really weird results, it might be worth confirming that you're pre-processing the testing data in the same exact way that you pre-processed the training data.

**Comment by ninjakx** [comment]
@JordanMakesMaps  I am not using 
```
test_dataset = Dataset(
    x_test_dir, 
    y_test_dir, 
    classes=CLASSES, 
    augmentation=get_validation_augmentation(),
    preprocessing=get_preprocessing(preprocess_input),
)
```
As it takes 2 inputs : input image and mask image. So I am not using that. Instead predicting like this : 
```
for each in os.listdir('test/'):
    print(each)
    img = cv2.imread(os.path.join('test', each))
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = preprocessing_fn(img)
    img = img.transpose(2, 0, 1)#.astype('float32')
    img = img.astype('uint8')/255.0
    x = torch.from_numpy(np.expand_dims(img, 0))
    y = model.predict(x.to(DEVICE))
```

**Comment by ninjakx** [comment]
where my x_test_dir is 'test/'

**Comment by ninjakx** [comment]
@JordanMakesMaps sorry I just overlooked it. Even if I remove  `img = img.astype('uint8')/255.0` problem remain the same.

**Comment by ninjakx** [comment]
```
class Dataset(BaseDataset):
    """
    # // TODO: Implement the Dataset class
    """
    
    CLASSES = ['tumour']
    
    def __init__(self, images_df, masks_df, class_values=[255], 
                 augmentation=None, preprocessing=None):
        """
        The custom dataset class that we use in pytorch
        to extract the images and their corresponding masks.
        
        All the augmentation and preprocessing functions should
        follow the following criteria
            a. They should accept image and mask
            b. They should return image and mask
            c. They should not change the dimension of image and mask
        
        :param image_dir: Path where the images are located
        :param masks_dir: Path where the corresponding masks are located
        :param class_values: The class values that we need to segment. For each
                             different object pass a different masks values.
                             By default this value is fixed for binary segmentation
        """
        # Getting all the images and their masks
        self.images_fps = []
        self.masks_fps = []
        # Performance of iterdir() is better
        for e in range(len(images_df)):
#             imgId = e.name
#             self.images_fps.append(str(e.resolve()))
#             self.masks_fps.append(str(masks_dir.joinpath(imgId).resolve()))
            self.images_fps.append(images_df[e])
            self.masks_fps.append(masks_df[e])
        
        # This indicates the class values that
        # we want to extract
        self.class_values = class_values
        
        # The augmentation we need to perform
        self.augmentation = augmentation
        
        # The image preprocessing that we need to perform
        self.preprocessing = preprocessing
    
    
    def __getitem__(self, i):
        """
        Return the ith image and it's corresponding
        mask values. This function defines how an
        image should be read and converted along with
        their masks.
        """
        # read data
        image = cv2.imread(self.images_fps[i])
        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        mask = cv2.imread(self.masks_fps[i], 0)
        mask[np.where(mask != 0)] = 255
#         print(np.unique(mask.ravel()))
#
             
        # extract classes from mask 
        masks = [(mask == v) for v in self.class_values]
        mask = np.stack(masks, axis=-1).astype('float32')
        

        
        # apply augmentations
        if self.augmentation:
            sample = self.augmentation(image=image, mask=mask)
            image, mask = sample['image'], sample['mask']
        
        # apply preprocessing
        if self.preprocessing:
            sample = self.preprocessing(image=image, mask=mask)
            
            # scaling the images
            image, mask = sample['image'], sample['mask']
            image = image.astype('float32')/255.0
            mask = mask.astype('float32')
        return image, mask
        
    def __len__(self):
        assert len(self.images_fps) == len(self.masks_fps)
        return len(self.images_fps)
```

The reason to use `img = img.astype('uint8')/255.0` is I have used preprocessing_fn and then divide it by 255 to make it converge fast. I am not sure what `preprocessing_fn` does.

**Comment by JordanMakesMaps** [comment]
Where is `preprocessing_fn` coming from? Is that an instance of `preprocess_input(backbone)`? Because if it is, you don't need to do `img = img.astype('uint8')/255.0`; @qubvel's `preprocess_input()` will do some normalization method on the images and is different for each specific backbone (VGG, ResNet, EfficientNet). 

So when you're making predictions on testing data, and you `preprocess_fn()` on the images and then rescaling them by 255, the the values are way out of wack (e.g. they're much smaller in value that what the training images were)

One way to check the values between each normalization process is using `numpy.ptp(img, axis = 2)`, which gives you the min-max range across channels. I think what's happening is you're rescaling the image twice?
